streaming/base/dataset.py

```diff
@@ -1,364 +1,384 @@
                  keep_zip: Optional[bool] = None,
-                 retry: int = 2,
-                 timeout: float = 60,
-                 hash: Optional[str] = None,
-                 batch_size: Optional[int] = None) -> None:
-        if keep_zip is None:
-            keep_zip = remote is None or remote == local
-        hash = hash or None
-
+                 download_retry: int = 2,
+                 download_timeout: float = 60,
+                 validate_hash: Optional[str] = None,
+                 shuffle_seed: Optional[int] = None,
+                 num_canonical_nodes: Optional[int] = None,
+                 batch_size: Optional[int] = None):
         self.local = local
         self.remote = remote
-        self.split = split or ''
+        self.split = split or ''  # Empty string for os.path.join().
         self.shuffle = shuffle
-        self.prefetch = prefetch
+        self.predownload = predownload
         self.keep_zip = keep_zip
-        self.retry = retry
-        self.timeout = timeout
-        self.hash = hash
+        self.download_retry = download_retry
+        self.download_timeout = download_timeout
+        self.validate_hash = validate_hash or None
+
+        if tdist.is_available() and torch.cuda.is_available() and not tdist.is_initialized():
+            tdist.init_process_group('nccl')
+
+        # Seed is set below.
+        world = World()
+        self.num_canonical_nodes = num_canonical_nodes
         self.batch_size = batch_size
+        self.shuffle_seed = shuffle_seed
 
+        # Load the index.json file.
         basename = get_index_basename()
-        wait = dist.get_local_rank() != 0
-        filename = self._download_file(basename, wait)
+        if world.is_local_leader:
+            filename = self._download_file(basename)
+        else:
+            filename = os.path.join(local, self.split, basename)  # pyright: ignore
+        dist.barrier()
         obj = json.load(open(filename))
-        assert obj['version'] == 2
+        if obj['version'] != 2:
+            raise ValueError('Unsupported version')
 
+        # Initialize shard readers according to the loaded info.
         self.shards = []
         for info in obj['shards']:
-            shard = reader_from_json(local, split, info)
+            shard = reader_from_json(local, self.split, info)
             self.shards.append(shard)
 
-        samples_per_shard = list(map(lambda shard: shard.samples, self.shards))
-        self.index = Index(samples_per_shard, batch_size)
+        # Build the Index (for partitioning and mapping samples to shards).
+        self.shard_sizes = np.array([x.samples for x in self.shards])
+        self.index = Index(self.shard_sizes)
+
+        # Determine and distribute shuffle seed and shm prefix.
+        if shuffle_seed is None:
+            shuffle_seed = np.random.randint(1 << 60)
+        prefix_int = np.random.randint(1 << 60)
+        if 1 < world.num_ranks:
```

Comment:
nit: slightly more consistent with code style elsewhere
```suggestion
        if world.num_ranks > 1:
```

Reply:
Done in https://github.com/mosaicml/streaming/pull/37/commits/82fb0ad0ae95bd0bc3d993605221a1bf928755c3.